### PR TITLE
Union merged

### PR DESCRIPTION
**Pull request by sushilmohanty**

### CODE REVIEW

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java [hunk 1]
```diff
@@ -13,58 +14,55 @@
 public class CandidateCoveringSetsResolver implements ContextRewriter {
 
   private List<Candidate> finalCandidates = new ArrayList<>();
-  private int unionCandidatealiasCounter = 0;
-  private int joinCandidatealiasCounter = 0;
+  //private int unionCandidatealiasCounter = 0;
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java]: Can we delete commented code ?

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java [hunk 2]
```diff
@@ -13,58 +14,55 @@
...
-    //if (cubeql.getCandidates().size() == 0) {
-    //  throw new LensException(LensCubeErrorCode.NO_FACT_HAS_COLUMN.getLensErrorInfo(), msrString);
-    // }
   }
 
-  private Candidate createJoinCandidateFromUnionCandidates(List<Candidate> ucs, CubeQueryContext cubeql) {
+  private Candidate createJoinCandidate(List<Candidate> ucs, CubeQueryContext cubeql) {
     Candidate cand;
-    if (ucs.size() >= 2) {
+   // if (ucs.size() >= 2) {
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java]: Delete commented code

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java [hunk 3]
```diff
@@ -13,58 +14,55 @@
...
-    cubeql.getCandidates().addAll(finalCandidates);
-    // TODO : we might need to prune if we maintian two data structures in CubeQueryContext.
-    //cubeql.pruneCandidateFactWithCandidateSet(CandidateTablePruneCause.columnNotFound(getColumns(queriedMsrs)));
-    //if (cubeql.getCandidates().size() == 0) {
-    //  throw new LensException(LensCubeErrorCode.NO_FACT_HAS_COLUMN.getLensErrorInfo(), msrString);
-    // }
   }
 
-  private Candidate createJoinCandidateFromUnionCandidates(List<Candidate> ucs, CubeQueryContext cubeql) {
+  private Candidate createJoinCandidate(List<Candidate> ucs, CubeQueryContext cubeql) {
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java]: can we rename ucs to childCandidates ?

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java [hunk 4]
```diff
@@ -74,7 +72,7 @@ private void updateFinalCandidates(List<List<Candidate>> jcs, CubeQueryContext c
       if (jc.size() == 1) {
         finalCandidates.add(jc.iterator().next());
       } else {
-        finalCandidates.add(createJoinCandidateFromUnionCandidates(jc, cubeql));
+        finalCandidates.add(createJoinCandidate(jc, cubeql));
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java]: Rename jcs and jc appropriately .

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateTableResolver.java [hunk 5]
```diff
@@ -424,10 +425,11 @@ private void resolveCandidateFactTablesForJoins(CubeQueryContext cubeql) throws
...
     Collection<String> colSet = null;
-    if (cubeql.getCube() != null && !cubeql.getCandidateFacts().isEmpty()) {
-      for (Iterator<CandidateFact> i = cubeql.getCandidateFacts().iterator(); i.hasNext();) {
-        CandidateFact cfact = i.next();
-        CubeFactTable fact = cfact.fact;
+    if (cubeql.getCube() != null && !cubeql.getCandidates().isEmpty()) {
+      for (Iterator<StorageCandidate> i =
+           CandidateUtil.getStorageCandidates(cubeql.getCandidates()).iterator(); i.hasNext();) {
+        StorageCandidate sc = i.next();
+        //CubeFactTable factTable =  i.next().getFact();
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateTableResolver.java]: Delete this comment or leave a TODO union

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CubeQueryContext.java [hunk 6]
```diff
@@ -949,21 +944,20 @@ private void addRangeClauses(StorageCandidate sc) throws LensException {
...
       // prune join paths for picked fact and dimensions
-      autoJoinCtx.pruneAllPaths(cube, cand, dimsToQuery);
+      autoJoinCtx.pruneAllPaths(cube, scSet, dimsToQuery);
     }
 
     Map<StorageCandidate, Set<Dimension>> factDimMap = new HashMap<>();
     if (cand != null) {
       // copy ASTs for each storage candidate
-      for (StorageCandidate sc : scSet) {
+      for (StorageCandidate sc : CandidateUtil.getStorageCandidates(cand)) {
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/CubeQueryContext.java]: Is this change required ? can we not use scSet ?

File: lens-cube/src/main/java/org/apache/lens/cube/parse/DefaultAliasDecider.java [hunk 7]
```diff
@@ -18,10 +18,14 @@
...
 package org.apache.lens.cube.parse;
 
+import lombok.Getter;
+import lombok.Setter;
 import org.apache.hadoop.hive.ql.parse.ASTNode;
 
 
 public class DefaultAliasDecider implements AliasDecider {
+  @Getter
+  @Setter
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/DefaultAliasDecider.java:28]: Setter is not needed

File: lens-cube/src/main/java/org/apache/lens/cube/parse/JoinCandidate.java [hunk 8]
```diff
@@ -28,29 +17,17 @@
...
-  private String alias;
   private QueryAST queryAST;
   private CubeQueryContext cubeql;
 
-  public JoinCandidate(Candidate childCandidate1, Candidate childCandidate2, String alias, CubeQueryContext cubeql) {
+  public JoinCandidate(Candidate childCandidate1, Candidate childCandidate2, CubeQueryContext cubeql) {
     this.childCandidate1 = childCandidate1;
     this.childCandidate2 = childCandidate2;
-    this.alias = alias;
+    //this.alias = alias;
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/JoinCandidate.java]: delete this line of leave a TODO union

**Review by guptapuneet — COMMENTED** [review]
Partial Review -1

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CubeQueryContext.java [hunk 9]
```diff
@@ -949,21 +944,20 @@ private void addRangeClauses(StorageCandidate sc) throws LensException {
...
         sc.setQueryAst(DefaultQueryAST.fromStorageCandidate(sc, this));
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/CubeQueryContext.java]: Can we do this instead . 
        QueryAST ast = DefaultQueryAST.fromStorageCandidate(sc, this);
        CandidateUtil.copyASTs(ast, sc.getQueryAst());

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 10]
```diff
@@ -54,24 +57,28 @@ public UnionQueryWriter(Candidate cand, CubeQueryContext cubeql) {
   public String toHQL() throws LensException {
     // Set the default value for the non queriable measures. If a measure is not
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java]: This comment is out of place , It should be above line updateInnterSelectASTWithDefault(); or better we can leave it as method comment for method updateInnterSelectASTWithDefault

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 11]
```diff
@@ -167,77 +193,152 @@ private ASTNode processOrderbyExpression(ASTNode astNode) throws LensException {
...
-        break;
+  private boolean isAggregateFunctionUsedInAST(ASTNode node) {
+      if (HQLParser.isAggregateAST(node)
+          || HQLParser.hasAggregate(node)) {
+        return true;
+      }
+    return false;
+  }
+
+  private void updateInnterSelectASTWithDefault() throws LensException {
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java]: Typo in method name

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 12]
```diff
@@ -167,77 +193,152 @@ private ASTNode processOrderbyExpression(ASTNode astNode) throws LensException {
...
+    for (int i = 0; i < cubeql.getSelectPhrases().size(); i++) {
+      SelectPhraseContext phrase = cubeql.getSelectPhrases().get(i);
+      ASTNode aliasNode = new ASTNode(new CommonToken(Identifier, phrase.getSelectAlias()));
+      if (!phrase.hasMeasures(cubeql)) {
+        for (StorageCandidate sc : storageCandidates) {
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java:271]: Please add comments for each if/else if/else  tp expalin what it intended.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 13]
```diff
@@ -167,77 +193,152 @@ private ASTNode processOrderbyExpression(ASTNode astNode) throws LensException {
...
-      // Check whether phrase is dimensions or answerable measure in StorageCandidate
-      if (!phrase.hasMeasures(cubeql)
-          || sc.getAnswerableMeasureIndices().contains(phrase.getPosition())) {
-        ASTNode node = getSelectExprAST();
-        node.addChild(getInnerSelectExprForAlias(sc, aliasNode.getText()));
-        node.addChild(aliasNode);
-        innerSelect.addChild(node);
-      } else {
-        innerSelect.addChild(getDefaultNode(aliasNode));
+  private ASTNode getAggregateNodesExpression(int position) {
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java:218]: can we leave detailed comments as to why we have chosen this approach to pick the expression from any one of the StorageCandidates and how we expect expression definition to be same across all StorageCandidates

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 14]
```diff
@@ -167,77 +193,152 @@ private ASTNode processOrderbyExpression(ASTNode astNode) throws LensException {
...
+    ASTNode node = null;
+    for (StorageCandidate sc : storageCandidates) {
+      node = (ASTNode) sc.getQueryAst().getSelectAST().getChild(position).getChild(0);
+      if (HQLParser.isAggregateAST(node) || HQLParser.hasAggregate(node)) {
+        return MetastoreUtil.copyAST(node);
+      }
+    }
+    return MetastoreUtil.copyAST(node);
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java:226]: Instead of return null can we throw AssertionError as this situation should never arise. 

  throw new AssertionError("Was expecting an aggregate expression but found no Storage Candidates " +
      "with aggregate Expression");

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 15]
```diff
@@ -167,77 +193,152 @@ private ASTNode processOrderbyExpression(ASTNode astNode) throws LensException {
...
+        }
+      } else if (!phrase.getQueriedMsrs().isEmpty()) {
+        for (StorageCandidate sc : storageCandidates) {
+          if (sc.getAnswerableMeasurePhraseIndices().contains(phrase.getPosition())) {
+            ASTNode exprWithOutAlias = (ASTNode) sc.getQueryAst().getSelectAST().getChild(i).getChild(0);
+            storageCandidateToSelectAstMap.get(sc.toString()).
+                addChild(getSelectExpr(exprWithOutAlias, aliasNode, false));
+          } else {
+            ASTNode resolvedExprNode = getAggregateNodesExpression(i);
+            if (isAggregateFunctionUsedInAST(resolvedExprNode)) {
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java:284]: isAggregateFunctionUsedInAST check looks redundant since same check is done inside getAggregateNodesExpression. Can we not check resolvedExprNode==null directly and decide ?

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 16]
```diff
@@ -167,77 +193,152 @@ private ASTNode processOrderbyExpression(ASTNode astNode) throws LensException {
...
+            } else {
+              resolvedExprNode = getSelectExpr(null, null, true);
+            }
+            storageCandidateToSelectAstMap.get(sc.toString()).
+                addChild(getSelectExpr(resolvedExprNode, aliasNode, false));
+          }
+        }
+      } else {
+        for (StorageCandidate sc : storageCandidates) {
+          if (phrase.isEvaluable(cubeql, sc)
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java:295]: Why phrase.isEvaluable(cubeql, sc) is required here ?   sc.getAnswerableMeasurePhraseIndices().contains(phrase.getPosition()) should be enough

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 17]
```diff
@@ -167,77 +193,152 @@ private ASTNode processOrderbyExpression(ASTNode astNode) throws LensException {
...
+    if (!sc.getColumns().containsAll(cols)) {
+      return true;
+    }
+  return false;
+  }
+
+  private ASTNode setDefaultValueInExprForAggregateNodes(ASTNode node, StorageCandidate sc)
+      throws LensException {
+    if (HQLParser.isAggregateAST(node)
+        && isNodeAnswerableForStorageCandidate(sc, node)) {
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java:241]: isNodeAnswerableForStorageCandidate() check should not be done .  Say there is an expression sum(m1)+sum(m2) and StorageCandidate has just m1 . In this case i expect 0.0 as m1, 0.0 as m2 to be projected. Looks like in this case Sum(m1), 0.0 as m2 will be projected which is wrong.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java (hunk 11 above)
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java]: "else if" and "else" parts have exactly same code . I think they can be combined. 
Can we combine the handling for else if = "sum(m1)"  and  else = "sum(m1) + sum(m2)"

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 18]
```diff
@@ -54,24 +57,28 @@ public UnionQueryWriter(Candidate cand, CubeQueryContext cubeql) {
   public String toHQL() throws LensException {
     // Set the default value for the non queriable measures. If a measure is not
     // answerable from a StorageCandidate set it as 0.0
-    for (StorageCandidate sc : storageCandidates) {
-      updateSelectASTWithDefault(sc);
-    }
+    updateAsts();
+    updateInnterSelectASTWithDefault();
+    // Set the default queryAST for the outer query
     queryAst = DefaultQueryAST.fromStorageCandidate(storageCandidates.iterator().next(),
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java]: Can we push this line above updateInnterSelectASTWithDefault(). 
Aslo can we not create iterator storageCandidates.iterator() twice .

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java [hunk 19]
```diff
@@ -167,77 +193,152 @@ private ASTNode processOrderbyExpression(ASTNode astNode) throws LensException {
...
     for (StorageCandidate sc : storageCandidates) {
+      aliasDecider.setCounter(0);
       ASTNode innerSelectAST = new ASTNode(new CommonToken(TOK_SELECT, "TOK_SELECT"));
-      innerToOuterASTs.clear();
-      AliasDecider aliasDecider = new DefaultAliasDecider();
-      ASTNode originalSelectAST = MetastoreUtil.copyAST(sc.getQueryAst().getSelectAST());
-      queryAst.setSelectAST(new ASTNode(originalSelectAST.getToken()));
       processSelectExpression(sc, outerSelectAst, innerSelectAST, aliasDecider);
-      outerHavingAst = processHavingAST(innerSelectAST, aliasDecider, sc);
+      selectAliasCounter = aliasDecider.getCounter();
```
Comment (guptapuneet) [inline lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java:323]: Can selectAliasCounter be set after for loop ?

**Review by guptapuneet — COMMENTED** [review]
Partial Review -2

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java (hunk 1 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java]: done.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java (hunk 3 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java]: done.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java (hunk 2 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java]: done.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CubeQueryContext.java (hunk 6 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/CubeQueryContext.java]: yes.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateTableResolver.java (hunk 5 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateTableResolver.java]: done.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/DefaultAliasDecider.java (hunk 7 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/DefaultAliasDecider.java:28]: Setter is used in UnionQueryRewrite to make alias count in sync.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/CubeQueryContext.java (hunk 9 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/CubeQueryContext.java]: +1.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java (hunk 18 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java]: done.

File: lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java (hunk 10 above)
Comment (sushilmohanty) [reply lens-cube/src/main/java/org/apache/lens/cube/parse/UnionQueryWriter.java]: done.
